File: day18/part1.exs
require IEx

defmodule InputHelpers do
  def parse(isTest) do
    filePath = if isTest, do: "test-data.txt", else: "data.txt"

    {:ok, rawData} = File.read(filePath)

    rawData
    |> String.split("\n")
    |> Enum.map(fn line ->
      String.split(line, ",", trim: true) |> Enum.map(&String.to_integer/1)
    end)
  end
end

defmodule Print do
  def print({index, movement, position, cols, rows}) do
    IO.puts("")
    IO.puts(movement)
    print_line(index, position, cols, rows, 0)
  end

  def print_line(index, position, cols, rows, row) do
    line = build_line(index, position, cols - 1, row, [])
    IO.puts(line)

    if row < rows do
      print_line(index, position, cols, rows, row + 1)
    end
  end

  def build_line(_, _, -1, _, line) do
    line |> Enum.join("")
  end

  def build_line(index, {posx, posy}, col, row, line) do
    build_line(index, {posx, posy}, col - 1, row, [
      if(col === posx && row === posy, do: "@", else: index[{col, row}]) | line
    ])
  end
end

defmodule Helpers do
  def dfs(stack, max, map, operation, history)

  def dfs([], _, _, _, history), do: history

  def dfs([{x, y} = key | stack], max, map, operation, history) do
    children =
      [{x, y + 1}, {x + 1, y}, {x, y - 1}, {x - 1, y}]
      |> Enum.filter(fn {x, y} ->
        x >= 0 && x <= max && y >= 0 && y <= max && map[{x, y}] !== true &&
          !Enum.member?(history, {x, y})
      end)

    next(&dfs/5, stack, key, children, max, map, operation, history)
  end

  defp next(named_function, collection, key, children, max, map, operation, history) do
    # IO.inspect(key)

    case operation.(max, key, history) do
      {:stop, res} ->
        [res | history]

      {:continue, res} ->
        children
        |> Enum.reduce(
          collection,
          fn tree, acc -> tree_insert(acc, tree) end
        )
        |> named_function.(max, map, operation, [res | history])
    end
  end

  defp tree_insert(collection, tree)
  defp tree_insert(collection, nil), do: collection
  defp tree_insert(stack, tree), do: [tree | stack]
end

defmodule Main do
  def run({isTest, bytes, max}) do
    map =
      InputHelpers.parse(isTest)
      |> Enum.take(bytes)
      |> Enum.reduce(%{}, fn [x, y], acc -> acc |> Map.put({x, y}, true) end)
      |> IO.inspect(label: "map")

    Helpers.dfs(
      [{0, 0}],
      max,
      map,
      fn max, {x, y}, _ ->
        if x === max && y === max, do: {:stop, {x, y}}, else: {:continue, {x, y}}
      end,
      []
    )
    |> IO.inspect(label: "path")
    |> length()
    |> IO.inspect(
      label:
        if isTest do
          "test result"
        else
          "real result"
        end
    )
  end
end

defmodule Test do
  use ExUnit.Case

  ExUnit.start()

  # @tag :skip
  test "Day 18 - Part 1 - test data" do
    assert Main.run({true, 12, 6}) == 22
  end

  @tag :skip
  test "Day 18 - Part 1 - real data" do
    assert Main.run({false, 1024, 70}) == 1_437_174
  end
end

File: day18/part1.mjs
import fs from 'fs'

const test = false

const inputTxt = fs.readFileSync(test ? './test-data.txt' : './data.txt', 'utf-8')

const toKey = ([x, y]) => `${x},${y}`

const bytes = test ? 12 : 1024
const obstacles = inputTxt
    .split(/\r?\n/)
    .slice(0, bytes)
    .reduce((result, coords) => {
        result[coords] = true
        return result
    }, {})
const max = test ? 6 : 70


const bfsSearch = (obstacles, max, start) => {
    const marked = {}
    const nodeFrom = {}

    const q = []
    marked[toKey(start)] = true
    q.push(start)

    while (q.length) {
        const [x, y] = q.shift()
        const cells = [
            [x, y + 1],
            [x - 1, y],
            [x + 1, y],
            [x, y - 1],
        ].filter(([x, y]) => x >= 0 && x <= max && y >= 0 && y <= max && !obstacles[toKey([x, y])])
        cells.forEach(c => {
            if (!marked[toKey(c)]) {
                nodeFrom[toKey(c)] = [x, y]
                marked[toKey(c)] = true
                q.push(c)
            }
        })
    }

    return nodeFrom
}

const pathTo = (nodeFrom, start, end) => {
    if (!nodeFrom[toKey(end)]) {
        throw new Error('Unreachable')
    }

    const path = []
    let node = end
    while (toKey(node) !== toKey(start)) {
        path.push(node)
        node = nodeFrom[node]
    }

    // path.push(start)
    path.reverse()
    return path
}

const print = (path, max) => {
    for (let y = 0; y <= max; y++) {
        const line = []
        for (let x = 0; x <= max; x++) {
            let cell = '.'
            if (obstacles[toKey([x, y])]) {
                cell = '#'
            }
            if (path.find(([x2, y2]) => x2 === x && y2 === y)) {
                cell = 'O'
            }
            line.push(cell)
        }
        console.log(line.join(''))
    }
}

const path = pathTo(bfsSearch(obstacles, max, [0, 0]), [0, 0], [max, max])
console.log(
    'result',
    path.length
)
// 316

// print(path, max)